source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/GenericImageLoaderAsync.java

```diff
@@ -0,0 +1,145 @@
+package io.adaptivecards.renderer;
+
+import android.content.Context;
+import android.content.res.Resources;
+import android.graphics.Bitmap;
+import android.graphics.BitmapFactory;
+import android.os.AsyncTask;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+
+import io.adaptivecards.renderer.http.HttpRequestHelper;
+import io.adaptivecards.renderer.http.HttpRequestResult;
+
+/** Abstract class that specifies image loading mechanism */
+public abstract class GenericImageLoaderAsync extends AsyncTask<String, Void, HttpRequestResult<Bitmap>>
+{
+    RenderedAdaptiveCard m_renderedCard;
+    String m_imageBaseUrl;
+
+    GenericImageLoaderAsync(RenderedAdaptiveCard renderedCard, String imageBaseUrl)
+    {
+        m_renderedCard = renderedCard;
+        m_imageBaseUrl = imageBaseUrl;
+    }
+
+    // Main function to try different ways to load an image
+    HttpRequestResult<Bitmap> loadImage(String path, Context context)
+    {
+        // Try loading online using only the path first
+        try
+        {
+            return loadOnlineImage(path);
+        }
+        catch (MalformedURLException e1) {
+            // Then try using image base URL to load online
+            try
+            {
+                if (m_imageBaseUrl == null || m_imageBaseUrl.isEmpty())
+                {
+                    return new HttpRequestResult<>(new Exception("Image base URL is empty or not specified"));
+                }
+
+                // Construct a URL using the image base URL and path
+                URL urlContext = new URL(m_imageBaseUrl);
+                URL url = new URL(urlContext, path);
+
+                return loadOnlineImage(url.toString());
+            }
+            catch (MalformedURLException e2)
+            {
+                // Then try reading it from local resources
+                return loadLocalImage(m_imageBaseUrl, context, path);
+            }
+            catch (Exception e)
+            {
+                return new HttpRequestResult<>(e);
+            }
+        }
+        catch (Exception excep)
+        {
+            return new HttpRequestResult<>(excep);
+        }
+    }
+
+    // Helper function to load online image URL
+    private HttpRequestResult<Bitmap> loadOnlineImage(String url) throws IOException, URISyntaxException
+    {
+        Bitmap bitmap;
```

Comment:
>Bitmap bitmap; [](start = 8, length = 14)

[nit] Move this declaration closer to the first use or do: 

Bitmap bitmap = BitmapFactory.decode ...